### ISSUE TITLE
[❔ other question] Is it possible to render a laser beam?

### ISSUE THREAD

**Issue by jma100** [mitsuba-renderer/mitsuba2#301]
## Summary

Is it possible to render a laser beam?

## Description

Is it possible to render a laser beam going through glass? If so, which emitter should I use? I tried distant directional emitter, which wasn't a single beam and spotlight seemed too wide for my purpose.

**Comment by Speierers** [comment]
Hi @jma100 ,

In order to render a laser beam in Mitsuba 2, you will need to implement a special emitter plugin that only emits light in a specific direction (for a specific point). Currently available plugins do not support this. You can take a look at the [`collimated`](https://github.com/mitsuba-renderer/mitsuba/blob/master/src/emitters/collimated.cpp) emitter plugin in Mitsuba 0.6.

On top of this, the existing `integrator` plugins (e.g. `path`) will not be able to render such images, as the sampling strategies used (next event estimation, BSDF sampling) won't be able to construct valid light path connecting with the laser emitter. Light paths will need to be constructed from the light source (here the laser). Such an integrator is also called "particle tracer" or "light tracer". 

Luckily for you, we are currently working on a implementation of a particle tracer in Mitsuba 2. I do not have any ETA for this feature unfortunately.

**Comment by jma100** [comment]
Thank you. Got it. Sorry I think I wasn't clear about exactly what we were looking for. What we actually wanted was to trace a ray starting from camera for a given pixel, entering and exiting a glass object, and then hitting a wall-like surface in the back. We want to know the position and direction of the ray hitting that wall-like surface. We don't actually need the rendering part.

Are there low-level libraries we can use to trace such ray and get the direction and position of the ray exiting the glass object?

**Comment by Speierers** [comment]
I would recommend you take a look at the Python API of Mitsuba 2. It should be fairly easy to implement such experiments with it.

This should be a good starting point: [Custom rendering pipeline in Python](https://mitsuba2.readthedocs.io/en/latest/src/python_interface/rendering_scene.html#custom-rendering-pipeline-in-python)

I will close this as it is not an issue. Please let me know if you have other questions

**Comment by jma100** [comment]
Sorry I'm still confused about how to do it. In the example from the link, it calls scene.ray_intersect(rays) and then get the depth. For the rays that are travelling through glass in our case, how do I keep marching the rays to get the final intersection with the wall (hitting glass, refracting, exiting the glass and hitting the wall)?

**Comment by Speierers** [comment]
Using the `SurfaceInteraction` data resulting from the ray intersection call, you should be able to refract the rays based on the properties of the glass material. You can then create a new wavefront of rays and call `scene.ray_intersect(new_rays)` once more. You could repeat this process many times until to reach the wall on the other side.

If you attach the proper dielectric BSDF to the glass object, you can also use the `BSDF::sample()` method to sample the next ray direction.

Basically you need to write your own custom integrator. Good luck!

**Comment by jma100** [comment]
Thank you for explaining. 

So from my understanding, the ray intersection doesn't take care of refraction. I need to get the refraction myself (using BSDF::sample()) and feed the refracted ray origin and directions into ray_intersect to do a new round of intersection. Is that right?

**Comment by jma100** [comment]
Hi, sorry I got another question. I tried to march the rays from camera to scene using ray_intersect and BSDF::sample(). The scene xml file I'm using contains one glass sphere only. Below are plots from different views of how a few camera rays travels (these are rays that remained active) after calling ray_intersect and BSDF::sample() 4 times:

light blue sphere = glass sphere
light blue line = optical axis
magenta lines = rays from pinhole to 1st intersection
red lines = rays from 1st to 2nd intersection
green lines = rays from 2nd to 3rd intersection
blue lines = rays from 3rd to 4th intersection

I guess what I was confused about was that I thought on the 3rd intersection (3rd time I'm calling ray_intersect), the red rays should exit the glass sphere and the active mask will become 0 for all rays, but instead they got reflected back into the glass sphere, kinda like internal reflection. How do I get the rays that go out of the glass sphere instead?

What my code looks like:
```
# pinhole to 1st intersection
si = scene.ray_intersect(rays)
ctx = BSDFContext()
bsdf = si.bsdf(rays)
active = si.is_valid()
active_b = active
bs, bsdf_val = BSDF.sample_vec(bsdf, ctx, si, sampler.next_1d(active), sampler.next_2d(active), active_b)

# 1st to 2nd intersection
ref_rays = si.spawn_ray(si.to_world(bs.wo))
si_bsdf = scene.ray_intersect(ref_rays, active_b)
bsdf_2nd = si_bsdf.bsdf(rays)
active_2nd = si_bsdf.is_valid()
active_b_2nd = active_2nd
bs_2nd, _ = BSDF.sample_vec(bsdf_2nd, ctx, si_bsdf, sampler.next_1d(active_2nd), sampler.next_2d(active_2nd), active_b_2nd)

# 2nd to 3rd intersection
ref_rays_2nd = si_bsdf.spawn_ray(si_bsdf.to_world(bs_2nd.wo))
si_bsdf_2nd = scene.ray_intersect(ref_rays_2nd, active_b_2nd)
active_3rd = si_bsdf_2nd.is_valid()
bsdf_3rd = si_bsdf_2nd.bsdf(rays)
bs_3rd, _ = BSDF.sample_vec(bsdf, ctx, si_bsdf_2nd, sampler.next_1d(active_3rd), sampler.next_2d(active_3rd), active_3rd)
active_b_3rd = active_3rd

# same thing for further intersection
```


![Screenshot from 2021-02-16 11-20-31](https://user-images.githubusercontent.com/22880685/108096380-e8c64500-704e-11eb-969d-f71a7dcd286e.png)
![Screenshot from 2021-02-16 11-20-50](https://user-images.githubusercontent.com/22880685/108096365-e532be00-704e-11eb-9ca5-1c073389ec1a.png)

**Comment by Speierers** [comment]
Hi @jma100 ,

I assume you are using a `dielectric` BSDF for your sphere in this scene? 

To answer your first question, the use of the random 2D point `sample2` will depend on the BSDF. E.g. the `dielectric` BSDF doesn't use it, while the `roughdielectric` BSDF use it to sample the normal distribution function.

I am not sure to understand what you are trying to achieve here. For instance, with the `dielectric` BSDF, the `sample1` random number will define whether to refract or reflect. In order to produce an un-biased estimation of the light transport equation, will need both at every bounce. Is this something specific to you integrator? Are you trying to limit the number of "internal" bounces
within the sphere, after which you always wanna connect to the camera? This is possible but would produce biased results. You can use the `BSDFContext` and `BSDFFlags` for this, enabling refraction/reflection only at a specific bounce.

**Comment by jma100** [comment]
Hi @Speierers, thanks again for your prompt reply.

Re: first question, yes I'm using a dielectric BSDF and got it.

Re: second question, here I'm only trying to get refractions and don't care about reflections. So for the rays in the image, I want to obtain the directions of where the red rays go when they exit the dielectric sphere. I guess based on what you mentioned, I just need to use the flag to enable refraction only for all bounces, like `self.m_flags = BSDFFlags.Transmission`. Is that right?

**Comment by Speierers** [comment]
You need to set the `type_mask` flag on the `BSDFContext` you pass to the sample call, e.g. 
```ctx.type_mask = BSDFFlags.DeltaTransmission```

Please take a look at `src/bsdfs/tests/test_dielectric.py` for more examples.
